RecoTracker/MkFit/plugins/MkFitOutputConverter.cc

```diff
@@ -0,0 +1,580 @@
+    const auto& param = state.parameters;
+    const auto& err = state.errors;
+    AlgebraicSymMatrix66 cov;
+    for (int i = 0; i < 6; ++i) {
+      for (int j = i; j < 6; ++j) {
+        cov[i][j] = err.At(i, j);
+      }
+    }
+
+    auto fts = FreeTrajectoryState(
+        GlobalTrajectoryParameters(
+            GlobalPoint(param[0], param[1], param[2]), GlobalVector(param[3], param[4], param[5]), state.charge, &mf),
+        CartesianTrajectoryError(cov));
+    if (!fts.curvilinearError().posDef()) {
+      edm::LogWarning("MkFitOutputConverter") << "Curvilinear error not pos-def\n"
+                                              << fts.curvilinearError().matrix() << "\noriginal 6x6 covariance matrix\n"
+                                              << cov << "\ncandidate ignored";
+      continue;
+    }
+
+    auto tsosDet =
+        backwardFitInCMSSW_
+            ? backwardFit(fts, recHits, propagatorAlong, propagatorOpposite, hitCloner, lastHitInvalid, lastHitChanged)
+            : convertInnermostState(fts, recHits, propagatorAlong, propagatorOpposite);
+    if (!tsosDet.first.isValid()) {
+      edm::LogWarning("MkFitOutputConverter")
+          << "Backward fit of candidate " << candIndex << " failed, ignoring the candidate";
+      continue;
+    }
+
+    // convert to persistent, from CkfTrackCandidateMakerBase
+    auto pstate = trajectoryStateTransform::persistentState(tsosDet.first, tsosDet.second->geographicalId().rawId());
+
+    output.emplace_back(
+        recHits,
+        seeds.at(seedIndex),
+        pstate,
+        seeds.refAt(seedIndex),
+        0,                                               // TODO: nloops, let's ignore for now
+        static_cast<uint8_t>(StopReason::UNINITIALIZED)  // TODO: ignore details of stopping reason as well for now
+    );
+  }
+  return output;
+}
+
+std::pair<TrajectoryStateOnSurface, const GeomDet*> MkFitOutputConverter::backwardFit(
+    const FreeTrajectoryState& fts,
+    const edm::OwnVector<TrackingRecHit>& hits,
+    const Propagator& propagatorAlong,
+    const Propagator& propagatorOpposite,
+    const TkClonerImpl& hitCloner,
+    bool lastHitWasInvalid,
+    bool lastHitWasChanged) const {
+  // First filter valid hits as in TransientInitialStateEstimator
+  TransientTrackingRecHit::ConstRecHitContainer firstHits;
+
+  for (int i = hits.size() - 1; i >= 0; --i) {
+    if (hits[i].det()) {
+      // TransientTrackingRecHit::ConstRecHitContainer has shared_ptr,
+      // and it is passed to backFitter below so it is really needed
+      // to keep the interface. Since we keep the ownership in hits,
+      // let's disable the deleter.
+      firstHits.emplace_back(&(hits[i]), edm::do_nothing_deleter{});
+    }
+  }
+
+  auto ret = backwardFitImpl(
+      fts, firstHits, propagatorAlong, propagatorOpposite, hitCloner, lastHitWasInvalid, lastHitWasChanged);
+  return ret;
+}
+
+std::pair<TrajectoryStateOnSurface, const GeomDet*> MkFitOutputConverter::backwardFitImpl(
```

Comment:
As the comments indicate, the implementation is *mostly* from [`RecoTracker/CkfPattern/src/TransientInitialStateEstimator.cc`](https://github.com/cms-sw/cmssw/blob/master/RecoTracker/CkfPattern/src/TransientInitialStateEstimator.cc). But here there is additional code *within* the copied lines, so `TransientInitialStateEstimator::innerState()` can not be called directly (in addition `TransientInitialStateEstimator::innerState()` takes `Trajectory`, which, I believe, we want to avoid here).

I'd prefer to leave the `backwardFit()` as it is and to be improved later if need arises. Alternatively I can remove the CMSSW-side backward fitting, because IIRC `mkFit`' backward fit worked much more reliably (in terms of producing non-ill tracks) and is hence the default.